shared/src/commonMain/kotlin/org/hyperskill/app/project_selection/presentation/ProjectSelectionListReducer.kt

```diff
@@ -0,0 +1,131 @@
+                    }
+                }
+            }
+            Message.RetryContentLoading -> {
+                state.doIf<ContentState.Error> {
+                    ContentState.Loading to
+                        fetchContent(state, forceLoadFromNetwork = true) +
+                            InternalAction.LogAnalyticEvent(
+                                ProjectsListClickedRetryContentLoadingHyperskillAnalyticsEvent(
+                                    trackId = state.trackId
+                                )
+                            )
+                }
+            }
+            is Message.ProjectClicked -> {
+                state.doIfContent { content ->
+                    content.copy(isProjectSelectionLoadingShowed = true) to
+                        setOf(
+                            InternalAction.SelectProject(
+                                trackId = state.trackId,
+                                projectId = message.projectId
+                            ),
+                            InternalAction.LogAnalyticEvent(
+                                ProjectsListClickedProjectHyperskillAnalyticsEvent(
+                                    trackId = state.trackId,
+                                    projectId = message.projectId
+                                )
+                            )
+                        )
+                }
+            }
+            is ProjectSelectionListFeature.ProjectSelectionResult -> {
+                state.doIfContent { content ->
+                    content.copy(isProjectSelectionLoadingShowed = false) to
+                        when (message) {
+                            ProjectSelectionListFeature.ProjectSelectionResult.Error ->
+                                setOf(ViewAction.ShowProjectSelectionStatus.Error)
+                            ProjectSelectionListFeature.ProjectSelectionResult.Success ->
+                                setOf(
+                                    ViewAction.NavigateTo.HomeScreen,
+                                    ViewAction.ShowProjectSelectionStatus.Success
+                                )
+                        }
+                }
+            }
+            Message.ViewedEventMessage -> {
+                state to setOf(
+                    InternalAction.LogAnalyticEvent(
+                        ProjectsListViewedHyperskillAnalyticEvent(state.trackId)
+                    )
+                )
+            }
+        }
+
+    private fun fetchContent(
+        state: State,
+        forceLoadFromNetwork: Boolean = false
+    ): Set<Action> =
+        setOf(
+            InternalAction.FetchContent(
+                state.trackId,
+                forceLoadFromNetwork = forceLoadFromNetwork
+            )
+        )
+
+    private fun State.updateContentState(contentState: ContentState): State =
+        copy(content = contentState)
+
+    private inline fun <reified T : ContentState> State.doIf(
```

Comment:
Could you please tell me what is the purpose of this extension? It removes some boilerplate code by checking `state.content` type?

Reply:
OK, I understood by looking at the `Message.RetryContentLoading`, but I'm not seeing benefits between writing `if ... else` and `doIf, doIfContent`.

In my opinion, it adds some kind of abstraction and complicates the reducer to be modified by the new developer, because all of our reducers are written in another style.

_Maybe I don't understand smth and you solved some problems with that extensions..._

Reply:
The state of the feature is not a sealed class, and it's not enough to use `if ... else`. In addition, you also should copy the state add update the `content`.
Without this extension, you have to write this boilerplate if you need content in `ContentState.Content` state.
```kotlin
if (content is ContentState.Content) {
    val (contentState, actions) = block(content)
    this.copy(content = contentState) to actions
} else {
    this to emptySet()
}
```
I think this tiny helper function is not so complicated and helps the reader concentrate on the logic.

Reply:
Actually, you are right, and `doIf` extension is not so needed, it reduces not so big amount of code.

Reply:
@XanderZhu you decided to remove those extensions?